File: scripts/analyze_tatum_payloads.py
#!/usr/bin/env python3
"""Analyze Tatum webhook payloads from SQL dumps."""
import json
import re
import sys
from collections import Counter
from pathlib import Path


def parse_sql_payloads(path: str) -> list[dict]:
    payloads: list[dict] = []
    text = Path(path).read_text(encoding="utf-8", errors="replace")
    for line in text.splitlines():
        line = line.strip()
        if not line.startswith("("):
            continue
        start = line.find('{"')
        if start == -1:
            start = line.find('{\\"')
        if start == -1:
            continue
        depth = 0
        end = -1
        for i in range(start, len(line)):
            ch = line[i]
            if ch == "{" and (i == start or line[i - 1] != "\\"):
                depth += 1
            elif ch == "}" and line[i - 1] != "\\":
                depth -= 1
                if depth == 0:
                    end = i + 1
                    break
        if end == -1:
            continue
        raw = line[start:end].replace('\\"', '"')
        try:
            payloads.append(json.loads(raw))
        except json.JSONDecodeError:
            pass
    return payloads


def normalize_payload(p: dict) -> dict:
    if isinstance(p.get("data"), dict):
        inner = dict(p["data"])
        for k in ("subscriptionType", "type", "chain", "subscriptionId"):
            if k not in inner and k in p:
                inner[k] = p[k]
        return inner
    return p


def summarize(name: str, payloads: list[dict]) -> None:
    print(f"\n=== {name} ({len(payloads)} payloads) ===")
    normalized = [normalize_payload(p) for p in payloads]

    sub_types = Counter(p.get("subscriptionType", "?") for p in normalized)
    print("subscriptionTypes:", dict(sub_types))

    types = Counter(str(p.get("type", "?")) for p in normalized)
    print("type field:", dict(types))

    currencies = Counter(str(p.get("currency", p.get("asset", "?"))) for p in normalized)
    print("currency/asset top:", currencies.most_common(20))

    chains = Counter(str(p.get("chain", "?")) for p in normalized)
    print("chain top:", chains.most_common(15))

    keys = Counter()
    for p in payloads:
        for k in p:
            keys[k] += 1
        if isinstance(p.get("data"), dict):
            for k in p["data"]:
                keys[f"data.{k}"] += 1

    print("top keys:", keys.most_common(30))

    fees = sum(
        1
        for p in normalized
        if str(p.get("type", "")).lower() == "fee"
        or (p.get("amount") is not None and str(p.get("amount")).startswith("-"))
    )
    print("fee/negative amount:", fees)

    no_sender = sum(
        1
        for p in normalized
        if not any(p.get(k) for k in ("counterAddress", "counteraddress", "from", "counter_address"))
    )
    print("no sender (BTC coinbase etc):", no_sender)

    has_account = sum(1 for p in normalized if p.get("accountId") or p.get("account_id"))
    print("has accountId:", has_account)

    has_asset = sum(1 for p in normalized if p.get("asset"))
    print("has asset field:", has_asset)

    has_contract = sum(1 for p in normalized if p.get("contractAddress"))
    print("has contractAddress:", has_contract)

    has_token_id = sum(1 for p in normalized if p.get("tokenId"))
    print("has tokenId:", has_token_id)

    has_to = sum(1 for p in normalized if p.get("to"))
    print("has to:", has_to)

    has_address_only = sum(1 for p in normalized if p.get("address") and not p.get("to"))
    print("address without to:", has_address_only)

    # Unique shape signatures
    shapes = Counter()
    for p in normalized:
        shape = (
            p.get("subscriptionType"),
            p.get("type"),
            "to" if p.get("to") else "addr" if p.get("address") else "noaddr",
            "asset" if p.get("asset") else "contract" if p.get("contractAddress") else "currency",
        )
        shapes[shape] += 1
    print("shape combos:", len(shapes))
    for shape, count in shapes.most_common(25):
        print(f"  {count:4d} {shape}")


def main() -> int:
    root = Path(__file__).resolve().parents[2]
    tatum_path = root / "tatum_raw_webhooks.sql"
    webhook_path = root / "webhook_raw_payloads.sql"

    tatum = parse_sql_payloads(str(tatum_path))
    webhook = parse_sql_payloads(str(webhook_path))

    summarize("tatum_raw_webhooks", tatum)
    summarize("webhook_raw_payloads", webhook)

    # Cross-check gaps vs current handler assumptions
    all_p = [normalize_payload(p) for p in tatum + webhook]
    unsupported_sub = [
        p.get("subscriptionType")
        for p in all_p
        if p.get("subscriptionType")
        not in (
            "ADDRESS_EVENT",
            "INCOMING_NATIVE_TX",
            "INCOMING_FUNGIBLE_TX",
            "ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION",
        )
    ]
    if unsupported_sub:
        print("\nUnsupported subscriptionTypes:", Counter(unsupported_sub))

    deposit_types = {"native", "token", None, "?"}
    skip_types = {"fee", "internal"}
    risky = [
        p
        for p in all_p
        if str(p.get("type", "")).lower() not in skip_types
        and p.get("subscriptionType")
        in (
            "ADDRESS_EVENT",
            "INCOMING_NATIVE_TX",
            "INCOMING_FUNGIBLE_TX",
            "ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION",
        )
        and float(str(p.get("amount", "0") or "0").replace("-", "") or 0) > 0
    ]
    print(f"\nTotal positive deposit-like payloads: {len(risky)}")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/analyze_tatum_payloads.php
<?php

/**
 * One-off analyzer for tatum_raw_webhooks.sql and webhook_raw_payloads.sql
 * Usage: php scripts/analyze_tatum_payloads.php
 */

function parseSqlPayloads(string $path): array
{
    $payloads = [];
    $lines = file($path, FILE_IGNORE_NEW_LINES);
    if ($lines === false) {
        return [];
    }

    foreach ($lines as $line) {
        $line = trim($line);
        if ($line === '' || ! str_starts_with($line, '(')) {
            continue;
        }

        $start = strpos($line, '{"');
        if ($start === false) {
            $start = strpos($line, '{\\"');
        }
        if ($start === false) {
            continue;
        }

        $depth = 0;
        $end = -1;
        $len = strlen($line);
        for ($i = $start; $i < $len; $i++) {
            $ch = $line[$i];
            if ($ch === '{' && ($i === $start || $line[$i - 1] !== '\\')) {
                $depth++;
            } elseif ($ch === '}' && $line[$i - 1] !== '\\') {
                $depth--;
                if ($depth === 0) {
                    $end = $i + 1;
                    break;
                }
            }
        }

        if ($end === -1) {
            continue;
        }

        $raw = str_replace('\\"', '"', substr($line, $start, $end - $start));
        $decoded = json_decode($raw, true);
        if (is_array($decoded)) {
            $payloads[] = $decoded;
        }
    }

    return $payloads;
}

function normalizePayload(array $p): array
{
    if (isset($p['data']) && is_array($p['data'])) {
        $inner = $p['data'];
        foreach (['subscriptionType', 'type', 'chain', 'subscriptionId'] as $k) {
            if (! array_key_exists($k, $inner) && array_key_exists($k, $p)) {
                $inner[$k] = $p[$k];
            }
        }

        return $inner;
    }

    return $p;
}

function summarize(string $name, array $payloads): void
{
    echo "\n=== {$name} (".count($payloads)." payloads) ===\n";

    $normalized = array_map('normalizePayload', $payloads);

    $subTypes = [];
    $types = [];
    $currencies = [];
    $chains = [];
    $keys = [];
    $shapes = [];

    $fees = 0;
    $noSender = 0;
    $hasAccount = 0;
    $hasAsset = 0;
    $hasContract = 0;
    $hasTokenId = 0;
    $hasTo = 0;
    $addressOnly = 0;

    foreach ($payloads as $p) {
        foreach (array_keys($p) as $k) {
            $keys[$k] = ($keys[$k] ?? 0) + 1;
        }
        if (isset($p['data']) && is_array($p['data'])) {
            foreach (array_keys($p['data']) as $k) {
                $keys['data.'.$k] = ($keys['data.'.$k] ?? 0) + 1;
            }
        }
    }

    foreach ($normalized as $p) {
        $st = (string) ($p['subscriptionType'] ?? '?');
        $subTypes[$st] = ($subTypes[$st] ?? 0) + 1;

        $t = (string) ($p['type'] ?? '?');
        $types[$t] = ($types[$t] ?? 0) + 1;

        $cur = (string) ($p['currency'] ?? $p['asset'] ?? '?');
        $currencies[$cur] = ($currencies[$cur] ?? 0) + 1;

        $chain = (string) ($p['chain'] ?? '?');
        $chains[$chain] = ($chains[$chain] ?? 0) + 1;

        $typeLower = strtolower($t);
        $amount = (string) ($p['amount'] ?? '');
        if ($typeLower === 'fee' || str_starts_with($amount, '-')) {
            $fees++;
        }

        $sender = $p['counterAddress'] ?? $p['counteraddress'] ?? $p['from'] ?? $p['counter_address'] ?? null;
        if ($sender === null || $sender === '') {
            $noSender++;
        }

        if (! empty($p['accountId']) || ! empty($p['account_id'])) {
            $hasAccount++;
        }
        if (! empty($p['asset'])) {
            $hasAsset++;
        }
        if (! empty($p['contractAddress'])) {
            $hasContract++;
        }
        if (! empty($p['tokenId'])) {
            $hasTokenId++;
        }
        if (! empty($p['to'])) {
            $hasTo++;
        }
        if (! empty($p['address']) && empty($p['to'])) {
            $addressOnly++;
        }

        $addrKind = ! empty($p['to']) ? 'to' : (! empty($p['address']) ? 'addr' : 'noaddr');
        $assetKind = ! empty($p['asset']) ? 'asset' : (! empty($p['contractAddress']) ? 'contract' : 'currency');
        $shape = $st.'|'.$t.'|'.$addrKind.'|'.$assetKind;
        $shapes[$shape] = ($shapes[$shape] ?? 0) + 1;
    }

    arsort($subTypes);
    arsort($types);
    arsort($currencies);
    arsort($chains);
    arsort($keys);
    arsort($shapes);

    echo 'subscriptionTypes: '.json_encode($subTypes)."\n";
    echo 'type field: '.json_encode($types)."\n";
    echo 'currency/asset top: '.json_encode(array_slice($currencies, 0, 20, true))."\n";
    echo 'chain top: '.json_encode(array_slice($chains, 0, 15, true))."\n";
    echo 'top keys: '.json_encode(array_slice($keys, 0, 30, true))."\n";
    echo "fee/negative amount: {$fees}\n";
    echo "no sender: {$noSender}\n";
    echo "has accountId: {$hasAccount}\n";
    echo "has asset: {$hasAsset}\n";
    echo "has contractAddress: {$hasContract}\n";
    echo "has tokenId: {$hasTokenId}\n";
    echo "has to: {$hasTo}\n";
    echo "address without to: {$addressOnly}\n";
    echo "shape combos: ".count($shapes)."\n";
    foreach (array_slice($shapes, 0, 25, true) as $shape => $count) {
        echo "  {$count} {$shape}\n";
    }
}

$root = dirname(__DIR__, 2);
$tatum = parseSqlPayloads($root.'/tatum_raw_webhooks.sql');
$webhook = parseSqlPayloads($root.'/webhook_raw_payloads.sql');

summarize('tatum_raw_webhooks', $tatum);
summarize('webhook_raw_payloads', $webhook);

$all = array_map('normalizePayload', array_merge($tatum, $webhook));
$unsupported = [];
$depositLike = 0;
foreach ($all as $p) {
    $st = (string) ($p['subscriptionType'] ?? '');
    if ($st !== '' && ! in_array($st, [
        'ADDRESS_EVENT',
        'INCOMING_NATIVE_TX',
        'INCOMING_FUNGIBLE_TX',
        'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION',
    ], true)) {
        $unsupported[$st] = ($unsupported[$st] ?? 0) + 1;
    }

    $typeLower = strtolower((string) ($p['type'] ?? ''));
    $amount = (float) str_replace('-', '', (string) ($p['amount'] ?? '0'));
    if (! in_array($typeLower, ['fee', 'internal'], true)
        && in_array($st, [
            'ADDRESS_EVENT',
            'INCOMING_NATIVE_TX',
            'INCOMING_FUNGIBLE_TX',
            'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION',
        ], true)
        && $amount > 0) {
        $depositLike++;
    }
}

echo "\nUnsupported subscriptionTypes: ".json_encode($unsupported)."\n";
echo "Total positive deposit-like payloads: {$depositLike}\n";

function inferSubscriptionType(array $data): string
{
    $subscriptionType = (string) ($data['subscriptionType'] ?? '');
    if ($subscriptionType !== '') {
        return $subscriptionType;
    }

    $kind = strtolower((string) ($data['kind'] ?? ''));
    if ($kind === 'token_transfer') {
        return 'INCOMING_FUNGIBLE_TX';
    }
    if (in_array($kind, ['native_transfer', 'native', 'transfer'], true)) {
        if (isset($data['tokenMetadata']['type']) && strtolower((string) $data['tokenMetadata']['type']) === 'fungible') {
            return 'INCOMING_FUNGIBLE_TX';
        }
        if (! empty($data['contractAddress']) || ! empty($data['tokenId'])) {
            return 'INCOMING_FUNGIBLE_TX';
        }

        return 'INCOMING_NATIVE_TX';
    }

    $payloadType = strtolower((string) ($data['type'] ?? ''));
    if (in_array($payloadType, ['token', 'native', 'fee'], true)) {
        return 'ADDRESS_EVENT';
    }

    if (! empty($data['contractAddress']) || ! empty($data['tokenId'])) {
        return 'INCOMING_FUNGIBLE_TX';
    }

    if (($data['amount'] ?? $data['value'] ?? '') !== '') {
        return 'INCOMING_NATIVE_TX';
    }

    return $payloadType;
}

function wouldSkipNonDeposit(array $data, string $subscriptionType): bool
{
    $payloadType = strtolower((string) ($data['type'] ?? ''));
    if (in_array($payloadType, ['fee', 'internal', 'failed'], true)) {
        return true;
    }

    $amount = (string) ($data['amount'] ?? '');
    if ($amount !== '' && str_starts_with($amount, '-')) {
        return true;
    }

    return str_starts_with($subscriptionType, 'OUTGOING_');
}

function allowsMissingSender(string $subscriptionType): bool
{
    return in_array($subscriptionType, [
        'INCOMING_NATIVE_TX',
        'INCOMING_FUNGIBLE_TX',
        'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION',
    ], true);
}

function classifyPayload(array $data): string
{
    $data = normalizePayload($data);
    $subscriptionType = inferSubscriptionType($data);

    if (wouldSkipNonDeposit($data, $subscriptionType)) {
        return 'ignored_non_deposit';
    }

    $txId = (string) ($data['txId'] ?? $data['txHash'] ?? $data['hash'] ?? '');
    if ($txId === '') {
        return 'failed_missing_tx';
    }

    $depositTypes = [
        'ADDRESS_EVENT',
        'INCOMING_NATIVE_TX',
        'INCOMING_FUNGIBLE_TX',
        'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION',
    ];
    if (! in_array($subscriptionType, $depositTypes, true)) {
        return 'failed_unsupported_type:'.$subscriptionType;
    }

    $to = $data['to'] ?? $data['address'] ?? null;
    if (! is_string($to) || trim($to) === '') {
        return 'failed_missing_address';
    }

    $sender = $data['counterAddress'] ?? $data['counteraddress'] ?? $data['from'] ?? $data['counter_address'] ?? null;
    if ($sender === null || $sender === '') {
        if ($subscriptionType === 'ADDRESS_EVENT') {
            return 'ignored_no_sender';
        }
        if (! allowsMissingSender($subscriptionType)) {
            return 'ignored_no_sender';
        }
    }

    return 'would_process_deposit';
}

$outcomes = [];
$depositWouldProcess = 0;
$depositWouldSkip = 0;
foreach ($all as $p) {
    $outcome = classifyPayload($p);
    $outcomes[$outcome] = ($outcomes[$outcome] ?? 0) + 1;
    if ($outcome === 'would_process_deposit') {
        $depositWouldProcess++;
    } elseif (! str_starts_with($outcome, 'ignored_non_deposit') && ! str_starts_with($outcome, 'failed_unsupported_type')) {
        $st = inferSubscriptionType(normalizePayload($p));
        if (in_array($st, [
            'ADDRESS_EVENT',
            'INCOMING_NATIVE_TX',
            'INCOMING_FUNGIBLE_TX',
            'ACCOUNT_INCOMING_BLOCKCHAIN_TRANSACTION',
        ], true) && ! wouldSkipNonDeposit(normalizePayload($p), $st)) {
            $depositWouldSkip++;
        }
    }
}

echo "\n=== Handler classification (no DB) ===\n";
ksort($outcomes);
foreach ($outcomes as $outcome => $count) {
    echo "  {$count} {$outcome}\n";
}
echo "Deposit-shaped that would reach address lookup: {$depositWouldProcess}\n";
echo "Deposit-shaped skipped before address lookup: {$depositWouldSkip}\n";
